### PR TITLE
Handle partition names like "loop0p1".

### PR DESCRIPTION
**Pull request by shishir-a412ed**
Signed-off-by: Shishir Mahajan <shishir.mahajan@redhat.com>
Signed-off-by: Marius Vollmer <mvollmer@redhat.com>

### CODE REVIEW

**Comment by shishir-a412ed** [comment]
@rhvgoyal The tests are passing locally on my system.

http://pastebin.test.redhat.com/468919
Need to look why it is failing on CI ?

@jlebon Any idea ?

Shishir

**Comment by shishir-a412ed** [comment]
````
Running test ./tests/001-test-use-devs-to-create-thin-pool.sh
INFO: Device node /dev/vdb1 exists.
wipefs: error: Checking: probing initialization failed: No such file or directory
ERROR: Failed to wipe signatures on device Checking that no-one is using this disk right now ... OK
````

Seems like the device `/dev/vdb` is already partitioned. Any idea why `/dev/vdb1` already exists ?

/cc @jlebon 

Shishir

File: container-storage-setup.sh [hunk 1]
```diff
@@ -731,40 +729,47 @@ create_partition_parted(){
...
-  local dev="$1"
+  local dev="$1" part
 
   if [ -x "/usr/sbin/parted" ]; then
     create_partition_parted $dev
   else
     create_partition_sfdisk $dev
   fi
 
+  part=$(query_first_child $dev)
```
Comment (mvollmer) [inline container-storage-setup.sh]: If this succeeds, then the partition exists and there is no point in calling `udevadm settle` or `wait_for_dev` below.  If you are worried that you need to wait for the partition block device to come into existence after calling `sfdisk`, then you should probably call `query_first_child` in a loop until it returns something non-empty.

However, the device for the partition does exist after calling `sfdisk` so there is no need to wait for it to appear.  What does happen, though, is that udevd asynchronously calls the BLKRRPART ioctl some time after the partition has been created, and because of the way BLKRRPART works, there is a brief period of time where the block device for the partition disappears.  If you want to protect against that, I recommend wrapping the whole `sfdisk; pvcreate; vgextend` operation in `flock -s $dev`.  The shared lock signals to udevd not to do the BLKRRPART ioctl.

Comment (mvollmer) [inline container-storage-setup.sh]: Also, if you think `udev settle` is needed, then you should call `query_first_child` after calling it, not before.

File: container-storage-setup.sh [hunk 2]
```diff
@@ -731,40 +729,47 @@ create_partition_parted(){
...
+  lsblk -npl -o NAME "$1" | tail -n +2 | head -1
 }
 
 create_disk_partitions() {
-  local devs="$1"
+  local devs="$1" part
 
   for dev in $devs; do
-    create_partition $dev
+    part=$(create_partition $dev)
```
Comment (mvollmer) [inline container-storage-setup.sh]: This will not work when `sfdisk` is used, I think.  `sfdisk` produces a lot of extra output on stdout.

**Comment by mvollmer** [comment]
What about removing all partition creation code from container-storage-setup?

File: container-storage-setup.sh (hunk 2 above)
Comment (mvollmer) [inline container-storage-setup.sh]: We can probably drop support for `sfdisk`, no?

**Comment by mvollmer** [comment]
```
wipefs: error: Checking: probing initialization failed: No such file or directory
ERROR: Failed to wipe signatures on device Checking that no-one is using this disk right now ... OK
```

This might be the issue I pointed out about `sfdisk` outputting a lot on stdout.  I think ${part} ends up being all of `Checking that no-one is using this disk right now ... OK....` and wipefs tries to use that as the device name.

File: container-storage-setup.sh (hunk 2 above)
Comment (shishir-a412ed) [inline container-storage-setup.sh]: We cannot drop support for `sfdisk`, that would break backward compatibility. Users who don't have `parted` will fall back to `sfdisk` for partitioning and we need to support that.

**Comment by shishir-a412ed** [comment]
@mvollmer Thanks for pointing it out. `sfdisk` output might be causing problems. Let me try to remove `parted` from my machine, then `container-storage-setup` would fallback to `sfdisk` and I can reproduce the issue.

**Comment by shishir-a412ed** [comment]
`What about removing all partition creation code from container-storage-setup?`

@mvollmer I will open a separate PR for that.

File: container-storage-setup.sh (hunk 1 above)
Comment (shishir-a412ed) [inline container-storage-setup.sh]: Okay makes sense. The `BLKRRPART` ioctl is a separate issue, I will drop that from the scope of this PR. @rhvgoyal Any specific reason you introduced `udevadm settle` and `wait_for_dev` ? Since if `create_partition_sfdisk` or `create_partition_parted` is successful, the device for the partition will exist. So we can just call `query_first_child()` and extract the partition name.

**Comment by shishir-a412ed** [comment]
@rhvgoyal PTAL.
I have rebase and resolved the conflicts against `upstream/master`, not sure why CI is not running through.

````
[smahajan@localhost docker-storage-setup]$ git branch
  docker_root_volume_test2
  dss_copy
  error_storage_driver
* handle_partition_names
  master
[smahajan@localhost docker-storage-setup]$ git pull --rebase upstream master
From github.com:projectatomic/docker-storage-setup
 * branch            master     -> FETCH_HEAD
Current branch handle_partition_names is up to date.
````

Shishir

File: container-storage-setup.sh [hunk 3]
```diff
@@ -755,26 +756,33 @@ create_partition() {
...
```
Comment (rhvgoyal) [inline container-storage-setup.sh]: how about querying partition after creation. Instead of returning it from create_partition. 

Anyway, create_partition() itself does not know better as it also is querying the partition name after creating it.

Downside is that we will query twice. But it removes the need of capturing the stdout of create_partition() and need of supressing sfdisk output. I like that better.

File: container-storage-setup.sh [hunk 4]
```diff
@@ -755,26 +756,33 @@ create_partition() {
...
+  part=$(query_first_child $dev)
+
+  if ! wait_for_dev ${part}; then
+    Fatal "Partition device ${part} is not available"
   fi
+  echo $part
+}
+
+query_first_child() {
+  lsblk -npl -o NAME "$1" | tail -n +2 | head -1
```
Comment (rhvgoyal) [inline container-storage-setup.sh]: How about doing 
`lsblk -npl -o NAME "$1" | head -n 2 | tail -1`

File: container-storage-setup.sh [hunk 5]
```diff
@@ -755,26 +756,33 @@ create_partition() {
...
-    Fatal "Partition device ${dev}1 is not available"
+  part=$(query_first_child $dev)
+
+  if ! wait_for_dev ${part}; then
+    Fatal "Partition device ${part} is not available"
   fi
+  echo $part
+}
+
+query_first_child() {
```
Comment (rhvgoyal) [inline container-storage-setup.sh]: lets call this dev_query_first_child()

File: container-storage-setup.sh [hunk 6]
```diff
@@ -721,18 +719,21 @@ scan_disks() {
...
-  local dev="$1" size
+  local dev="$1" size output
   # Use a single partition of a whole device
   # TODO:
   #   * Consider gpt, or unpartitioned volumes
   #   * Error handling when partition(s) already exist
   #   * Deal with loop/nbd device names. See growpart code
   size=$(( $( awk "\$4 ~ /"$( basename $dev )"/ { print \$3 }" /proc/partitions ) * 2 - 2048 ))
-    cat <<EOF | sfdisk $dev
+  if ! output=$(cat <<EOF | sfdisk $dev
```
Comment (rhvgoyal) [inline container-storage-setup.sh]: we can get rid of capturing output in a variable and supressing it once you stop returning partition name from create_partition()

File: container-storage-setup.sh (hunk 3 above)
Comment (shishir-a412ed) [inline container-storage-setup.sh]: Makes sense. Fixed.

File: container-storage-setup.sh (hunk 5 above)
Comment (shishir-a412ed) [inline container-storage-setup.sh]: Fixed.

File: container-storage-setup.sh (hunk 6 above)
Comment (shishir-a412ed) [inline container-storage-setup.sh]: Fixed.

**Comment by shishir-a412ed** [comment]
@rhvgoyal PTAL.

File: container-storage-setup.sh [hunk 7]
```diff
@@ -755,26 +753,33 @@ create_partition() {
...
-    Fatal "Partition device ${dev}1 is not available"
+  part=$(dev_query_first_child $dev)
+
+  if ! wait_for_dev ${part}; then
+    Fatal "Partition device ${part} is not available"
   fi
 }
 
+dev_query_first_child() {
+  lsblk -npl -o NAME "$1" | tail -n +2 | head -1
```
Comment (rhvgoyal) [inline container-storage-setup.sh]: I had given a comment on this to simplify it. can you please take care of that as well.

Comment (shishir-a412ed) [inline container-storage-setup.sh]: @rhvgoyal https://github.com/projectatomic/container-storage-setup/pull/234#discussion_r109312810

File: container-storage-setup.sh (hunk 4 above)
Comment (rhvgoyal) [inline container-storage-setup.sh]: I was thinking that it is easier to understand. If we are querying first child, then take first two lines of output. That will have parent and first child. 

tail -n +2, will return you all children. So I think it is simpler to do just first two lines?

Comment (shishir-a412ed) [inline container-storage-setup.sh]: Both of them seems exactly the same to me :). `tail -n +2` will return everything except the device and then you take the first partition out of it. e.g.
````
/dev/vdb
/dev/vdb1
/dev/vdb2
/dev/vdb3.
````
`tail` command will return everything except the device (`/dev/vdb`) and then you pick up the first partition using `head` which is `/dev/vdb1`. If your format seems simpler to you, I will change it.

**Comment by shishir-a412ed** [comment]
@rhvgoyal PTAL.

**Comment by shishir-a412ed** [comment]
@rhvgoyal Actually the tests are failing on this new format
https://github.com/projectatomic/container-storage-setup/pull/234#discussion_r109229653
in `dev_query_first_child()` so we need to go back to the original command which was working.

File: container-storage-setup.sh (hunk 4 above)
Comment (rhvgoyal) [inline container-storage-setup.sh]: ok, we can leave it as it is. I think I was much more familiar with head -2. For tail -n +2, I had to read the man page and that's one reason I thought it was simpler. 

No big deal. I will test it.

**Comment by rhvgoyal** [comment]
@shishir-a412ed one more minor nit. Try passing in /dev/loop0p1 as input and see what happens.

`DEVS=/dev/loop0p1`

lsblk does not report it as partitioned device. So we might have to do something additional. May be for device type loop, we can check if partition name ends with p*. Or try to do something by scanning /proc/partitions.

**Comment by shishir-a412ed** [comment]
@rhvgoyal Right. the `type` of a loop block device partition is `loop` and not `part`. As an experiment I partitioned a loop block device (`/dev/loop0`) and a regular VM Disk (`/dev/vdb`) on my system using `parted`.

````
sh-4.3# lsblk
NAME            MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
sr0              11:0    1 1024M  0 rom  
vda             252:0    0   40G  0 disk 
├─vda1          252:1    0  500M  0 part /boot
└─vda2          252:2    0 39.5G  0 part 
  ├─fedora-root 253:0    0 37.5G  0 lvm  /
  └─fedora-swap 253:1    0    2G  0 lvm  [SWAP]
vdb             252:16   0   20G  0 disk 
└─vdb1          252:17   0   20G  0 part 
loop0             7:0    0    5G  0 loop 
└─loop0p1       259:0    0    5G  0 loop 
````

As you can see `/dev/vdb1` has type `part` and `/dev/loop0p1` has type `loop`.
Right now there is already a check in `is_block_dev_partition()` to check if the block device was already partitioned or not.
````
  if [ "$disktype" == "part" ];then
    return 0
  fi
 ````
We can add a similar check to handle loop block devices partition. Something like
````
If [ "$disktype == "loop" && (Partition name ends with p*) ];then
   return 0
fi
````

**Comment by shishir-a412ed** [comment]
@rhvgoyal PTAL.

File: container-storage-setup.sh [hunk 8]
```diff
@@ -624,6 +624,15 @@ is_block_dev_partition() {
     return 0
   fi
 
+  # This will check if a loop/nvme device is passed and is already partitioned.
```
Comment (rhvgoyal) [inline container-storage-setup.sh]: Lets mention something like.

` For loop device partitions, lsblk reports type as "loop" and not "part". So check if device has a parent in the tree and if it does, there are high chances it is partition (except the case of lvm volumes)"

File: container-storage-setup.sh [hunk 9]
```diff
@@ -624,6 +624,15 @@ is_block_dev_partition() {
...
+  if ! devicePartition=$(lsblk -npls -o NAME ${bdev}|tail -n +2); then
```
Comment (rhvgoyal) [inline container-storage-setup.sh]: how about calling it "devparent". And also make it `local` variable to funciton.

File: container-storage-setup.sh (hunk 8 above)
Comment (shishir-a412ed) [inline container-storage-setup.sh]: Fixed.

File: container-storage-setup.sh (hunk 9 above)
Comment (shishir-a412ed) [inline container-storage-setup.sh]: Fixed.

**Comment by rhvgoyal** [comment]
Give me few minutes to finish review before you push new patch, please.

**Comment by shishir-a412ed** [comment]
@rhvgoyal ack.

**Comment by rhvgoyal** [comment]
@shishir-a412ed seeing some strange behavior. Try this.

- Create loop device
- Create partition
- Delete partition
- Use DEVS=/dev/loop0 and run container-storage-setup

I get following.

[root@vm7-f25 rhvgoyal-docker-storage-setup]# ./container-storage-setup.sh 
INFO: Device node /dev/loop0p1 exists.
wipefs: /dev/loop0p1: ignoring nested "dos" partition table on non-whole disk device
wipefs: /dev/loop0p1: ignoring nested "PMBR" partition table on non-whole disk device
wipefs: Use the --force option to force erase.
WARNING: dos signature detected on /dev/loop0p1 at offset 510. Wipe it? [y/n]:

**Comment by rhvgoyal** [comment]
Can we also add two tests.

- One test to make sure this works with loop devices
- Second test to make sure passing in partition loop device fails.

**Comment by shishir-a412ed** [comment]
@rhvgoyal PTAL.

File: container-storage-setup.sh [hunk 10]
```diff
@@ -755,26 +764,33 @@ create_partition() {
...
     create_partition $dev
+    part=$(dev_query_first_child $dev)
 
     # By now we have ownership of disk and we have checked there are no
     # signatures on disk or signatures should be wiped. Don't care
     # about any signatures found in the middle of disk after creating
     # partition and wipe signatures if any are found.
-    if ! wipefs -a ${dev}1; then
-      Fatal "Failed to wipe signatures on device ${dev}1"
+    if ! wipefs -f -a ${part}; then
```
Comment (rhvgoyal) [inline container-storage-setup.sh]: Lets keep this in a separate patch.

Comment (shishir-a412ed) [inline container-storage-setup.sh]: Fixed.

File: container-storage-setup.sh [hunk 11]
```diff
@@ -770,7 +770,7 @@ create_disk_partitions() {
     # signatures on disk or signatures should be wiped. Don't care
     # about any signatures found in the middle of disk after creating
     # partition and wipe signatures if any are found.
-    if ! wipefs -a ${dev}1; then
+    if ! wipefs -f -a ${dev}1; then
```
Comment (rhvgoyal) [inline container-storage-setup.sh]: We need to improve changelog. How about following.

`After partition creation, we wipe any signatures on the partition using wipefs. For some reason, for loop devices it was detecting nested partition table signature and wipefs will skip it by default and instead ask user whether to clean it or not. 

Use option -f instead to not ask user the question and delete it alaways.
`

Comment (shishir-a412ed) [inline container-storage-setup.sh]: s/alaways/always.

Comment (shishir-a412ed) [inline container-storage-setup.sh]: Fixed.

**Comment by shishir-a412ed** [comment]
@rhvgoyal PTAL.

File: tests/015-test-fail-loop-to-create-thin-pool.sh [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    return $test_status
+  fi
+
+  # Create loopback device.
+   tmpfile=$(mktemp /tmp/c-s-s.XXXXXX)
+   truncate --size=6G $tmpfile
+   lbdevice=$(losetup -f)
+   losetup --partscan $lbdevice $tmpfile
+   if ! create_partition $lbdevice;then
+      echo "ERROR: Failed partitioning $lbdevice"
```
Comment (rhvgoyal) [inline tests/015-test-fail-loop-to-create-thin-pool.sh]: If we fail to partition the device, we should fail test and return with status 1 here, isn't it?

File: tests/115-test-fail-loop-to-create-thin-pool.sh [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    return $test_status
+  fi
+
+  # Create loopback device.
+   tmpfile=$(mktemp /tmp/c-s-s.XXXXXX)
+   truncate --size=6G $tmpfile
+   lbdevice=$(losetup -f)
+   losetup --partscan $lbdevice $tmpfile
+   if ! create_partition $lbdevice;then
+      echo "ERROR: Failed partitioning $lbdevice"
```
Comment (rhvgoyal) [inline tests/115-test-fail-loop-to-create-thin-pool.sh]: Same here, if we fail to partition device, we should fail test and return 1 here.

File: tests/115-test-fail-loop-to-create-thin-pool.sh [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+  cleanup $vg_name "$lbdevice" "$infile" "$outfile"
+  cleanup_loop_device "$tmpfile" "$lbdevice"
+  return $test_status
+}
+
+cleanup_loop_device() {
+  local tmpfile=$1
+  local lbdevice=$2
+  rm $tmpfile > /dev/null 2>&1
+  losetup -d $lbdevice
```
Comment (rhvgoyal) [inline tests/115-test-fail-loop-to-create-thin-pool.sh]: This applies to all the four places you have used this in tests.

**Comment by rhvgoyal** [comment]
@shishir-a412ed Patches look more or less good to me. Gave some comments just now. Can you please take care of these and then I should be able to merge patches.

File: tests/115-test-fail-loop-to-create-thin-pool.sh (hunk 14 above)
Comment (shishir-a412ed) [inline tests/115-test-fail-loop-to-create-thin-pool.sh]: Fixed.

File: tests/115-test-fail-loop-to-create-thin-pool.sh (hunk 13 above)
Comment (shishir-a412ed) [inline tests/115-test-fail-loop-to-create-thin-pool.sh]: Fixed.

File: tests/015-test-fail-loop-to-create-thin-pool.sh (hunk 12 above)
Comment (shishir-a412ed) [inline tests/015-test-fail-loop-to-create-thin-pool.sh]: Fixed.

**Comment by shishir-a412ed** [comment]
@rhvgoyal PTAL.

**Comment by rhvgoyal** [comment]
LGTM. Thanks.

**Comment by rhvgoyal** [comment]
@rh-atomic-bot r+ 7ce2091

**Comment by rhvgoyal** [comment]
@rh-atomic-bot r+ 7ce2091
